pypiserver/core.py

```diff
@@ -1,154 +1,68 @@
-    elif path.count("-") == 1:
-        pkgname, version = path.split("-", 1)
-    elif "." not in path:
-        pkgname, version = path.rsplit("-", 1)
-    else:
-        pkgname = _pkgname_re.split(path)[0]
-        ver_spec = path[len(pkgname) + 1 :]
-        parts = _pkgname_parts_re.split(ver_spec)
-        version = parts[0]
-    return pkgname, version
-
-
-def normalize_pkgname(name):
-    """Perform PEP 503 normalization"""
-    return re.sub(r"[-_.]+", "-", name).lower()
-
-
-def normalize_pkgname_for_url(name):
-    """Perform PEP 503 normalization and ensure the value is safe for URLs."""
-    return quote(re.sub(r"[-_.]+", "-", name).lower())
-
-
-def is_allowed_path(path_part):
-    p = path_part.replace("\\", "/")
-    return not (p.startswith(".") or "/." in p)
+def get_bad_url_redirect_path(request, project):
+    """Get the path for a bad root url."""
+    uri = request.custom_fullpath
+    if uri.endswith("/"):
+        uri = uri[:-1]
+    uri = uri.rsplit("/", 1)[0]
+    project = quote(project)
+    uri += f"/simple/{project}/"
+    return uri
 
 
 class PkgFile:
-
     __slots__ = [
-        "fn",
-        "root",
-        "_fname_and_hash",
-        "relfn",
-        "relfn_unix",
-        "pkgname_norm",
-        "pkgname",
-        "version",
-        "parsed_version",
-        "replaces",
+        "pkgname",  # The projects/package name with possible capitalization
+        "version",  # The package version as a string
+        "fn",  # The full file path
+        "root",  # An optional root directory of the file
+        "relfn",  # The file path relative to the root
+        "replaces",  # The previous version of the package (used by manage.py)
+        "pkgname_norm",  # The PEP503 normalized project name
+        "digest",  # The file digest in the form of <algo>=<hash>
+        "relfn_unix",  # The relative file path in unix notation
+        "parsed_version",  # The package version as a tuple of parts
+        "digester",  # a function that calculates the digest for the package
     ]
+    digest: t.Optional[str]
+    digester: t.Optional[t.Callable[["PkgFile"], t.Optional[str]]]
+    parsed_version: tuple
+    relfn_unix: t.Optional[str]
 
     def __init__(
-        self, pkgname, version, fn=None, root=None, relfn=None, replaces=None
+        self,
+        pkgname: str,
+        version: str,
+        fn: t.Optional[str] = None,
+        root: t.Optional[str] = None,
+        relfn: t.Optional[str] = None,
+        replaces: t.Optional["PkgFile"] = None,
     ):
         self.pkgname = pkgname
         self.pkgname_norm = normalize_pkgname(pkgname)
         self.version = version
-        self.parsed_version = parse_version(version)
+        self.parsed_version: tuple = parse_version(version)
         self.fn = fn
         self.root = root
         self.relfn = relfn
         self.relfn_unix = None if relfn is None else relfn.replace("\\", "/")
         self.replaces = replaces
+        self.digest = None
```

Comment:
Should this be `digester` to match the `__slots__`?

Reply:
Both digest and digester are part of slots. I do see that I need to assign `self.digester=None` in `__init__` to be consistent